src/main/clojure/cemerick/pomegranate/aether.clj

```diff
@@ -797,28 +797,34 @@ kwarg to the repository kwarg.
         coordinates (merge-versions-from-managed-coords coordinates managed-coordinates)
         deps (coords->Dependencies files coordinates)
         managed-deps (coords->Dependencies files managed-coordinates)
-        collect-request (doto (CollectRequest. deps
-                                               managed-deps
+        collect-request (doto (CollectRequest. ^java.util.List deps
+                                               ^java.util.List managed-deps
+                                               ^java.util.List
                                                (vec (map #(let [repo (make-repository % proxy)]
-                                                            (-> session
+                                                            (-> ^RepositorySystemSession session
                                                                 (.getMirrorSelector)
                                                                 (.getMirror repo)
                                                                 (or repo)))
                                                          repositories)))
                           (.setRequestContext "runtime"))]
     (if retrieve
-      (.resolveDependencies system session (DependencyRequest. collect-request nil))
-      (.collectDependencies system session collect-request))))
+      (.resolveDependencies ^RepositorySystem system session (DependencyRequest. collect-request nil))
+      (.collectDependencies ^RepositorySystem system session collect-request))))
 
 (defn resolve-dependencies
   "Same as `resolve-dependencies*`, but returns a graph of dependencies; each
    dependency's metadata contains the source Aether Dependency object, and
    the dependency's :file on disk.  Please refer to `resolve-dependencies*` for details
    on usage, or use it if you need access to Aether dependency resolution objects."
   [& args]
-  (-> (apply resolve-dependencies* args)
-    .getRoot
-    dependency-graph))
+  (let [result (apply resolve-dependencies* args)]
+    (if (instance? DependencyResult result)
```

Comment:
I have some reservations about such changes. This is not just type-hinting, it’s changing the program logic. Subtle, true, but you’re unrolling the dynamic dispatch into manual dispatch, and so you’re potentially changing the behaviour of the code when something changes in the Maven resolver implementation.

Therefore I would advise against such changes (unless you are an expert on the resolver API and can guarantee that this change is sound, now and in the future).

Reply:
Spoke too early, this doesn't seem to apply here. Please ignore my comments.

Reply:
I wasn't sure about the best way to handle this, as `resolve-dependencies*` can return one of two classes that, although they both have a `.getRoot` method, are otherwise unrelated, so there's not a single type that works as a type hint.
Another option would have been to create a protocol and extend it to the two classes, but that seemed like overkill.

Reply:
Just to finish the thought of the OP (it wasn't obvious to me), this concern probably doesn't apply because we control which classes can reach this code via `resolve-dependencies*`.